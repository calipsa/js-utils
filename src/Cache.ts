import Singleton from './Singleton'

export default class Cache<K, V> extends Singleton {
  map = new Map<K, V>()

  add(idx: K, item: V) {
    this.map.set(idx, item)
  }

  get(idx: K) {
    return this.map.get(idx)
  }

  clear() {
    this.map.clear()
  }
}
