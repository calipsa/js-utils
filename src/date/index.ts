import fromNow from './fromNow'

export {
  // eslint-disable-next-line import/prefer-default-export
  fromNow,
}
