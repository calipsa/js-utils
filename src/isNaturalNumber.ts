/**
 * Checks if a value is a natural number,
 * i.e. is an integer greater than zero
 */
export default (num: any): num is number =>
  Number.isInteger(num) && num > 0
