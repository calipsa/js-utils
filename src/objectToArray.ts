import isNonNegativeInteger from './isNonNegativeInteger'

type O = Readonly<{
  [k: number]: any,
}>

/**
 * Takes all values matching non-negative integer keys in an object
 * and puts them in an array
 */
export default <Obj extends O, V extends Obj[keyof Obj]>(o: Obj) => {
  const indices = Object.keys(o)
    .map(Number)
    .filter(isNonNegativeInteger)

  const minKey = Math.min(...indices)
  const maxKey = Math.max(...indices)
  const arr: V[] = []

  for (let i = minKey; i <= maxKey; ++i) {
    arr[i] = o[i]
  }

  return arr
}
