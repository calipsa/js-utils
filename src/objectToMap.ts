/**
 * Converts an object to a `Map`
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export default <T extends {}, K extends keyof T>(o: T) =>
  new Map<string, T[K]>(Object.entries(o))
