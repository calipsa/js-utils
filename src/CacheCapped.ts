import Cache from './Cache'

export default class CacheCapped<K, V> extends Cache<K, V> {
  private capacity: number

  constructor(capacity: number) {
    super()
    this.capacity = capacity
  }

  rm(item: K) {
    this.map.delete(item)
  }

  add(idx: K, item: V) {
    if (this.map.size === this.capacity && !this.map.has(idx)) {
      const oldest = this.map.keys().next().value
      this.rm(oldest)
    }
    this.map.set(idx, item)
  }
}
