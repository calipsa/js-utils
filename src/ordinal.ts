import isNaturalNumber from './isNaturalNumber'

const got = (i: any) =>
  typeof i === 'string' ? JSON.stringify(i) : i

export default (i: number) => {
  if (!isNaturalNumber(i)) {
    throw new Error(`Number must be a natural number. Instead got: ${got(i)}`)
  }

  const j = i % 10
  const k = (i % 100) - j
  if (k !== 10) {
    if (j === 1) {
      return `${i}st`
    }
    if (j === 2) {
      return `${i}nd`
    }
    if (j === 3) {
      return `${i}rd`
    }
  }
  return `${i}th`
}
