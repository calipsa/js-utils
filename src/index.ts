import * as date from './date'
import * as timezone from './timezone'

import Cache from './Cache'
import CacheCapped from './CacheCapped'
import delay from './delay'
import hasDuplicates from './hasDuplicates'
import isNonNegativeInteger from './isNonNegativeInteger'
import objectToArray from './objectToArray'
import objectToFunction from './objectToFunction'
import objectToFunctionWithBind from './objectToFunctionWithBind'
import queryObjectToString from './queryObjectToString'
import Singleton from './Singleton'
import sliceRotate from './sliceRotate'
import toNearestByStep from './toNearestByStep'

export {
  date,
  timezone,

  Cache,
  CacheCapped,
  delay,
  hasDuplicates,
  isNonNegativeInteger,
  objectToArray,
  objectToFunction,
  objectToFunctionWithBind,
  queryObjectToString,
  Singleton,
  sliceRotate,
  toNearestByStep,
}
