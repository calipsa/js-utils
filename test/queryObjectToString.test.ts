import { queryObjectToString } from '../src'

describe('sliceRotate', () => {
  it('should convert an empty object correctly', () => {
    const o = {}
    const result = queryObjectToString(o)
    expect(result).toMatchSnapshot()
  })

  it('should convert a normal object correctly', () => {
    const o = {
      a: 5,
      b: 'foo',
      tr: true,
      fs: false,
      trs: 'true',
      fss: 'false',
      emptystring: '',
      d: new Date('2019-11-22T22:17:17.771Z'),
      ud: undefined,
    }
    const result = queryObjectToString(o)
    expect(result).toMatchSnapshot()
  })

  it('should convert an object with various cases correctly', () => {
    const o = {
      snake_case: 'val',
      camelCase: 'val',
      Capital: 'val',
      ALL_CAPS: 'val',
    }
    const result = queryObjectToString(o)
    expect(result).toMatchSnapshot()
  })

  it('should convert an object with numeric keys correctly', () => {
    const o = {
      11: 'eleven',
      0: 'zero',
    }
    const result = queryObjectToString(o)
    expect(result).toMatchSnapshot()
  })

  it('should convert a non-english object correctly', () => {
    const o = {
      ключ: 'значение',
      anahtar: 'değer',
      Schlüssel: 'Werte',
    }
    const result = queryObjectToString(o)
    expect(result).toMatchSnapshot()
  })

  it('should convert a weird object correctly', () => {
    const o = {
      twowords: 'kaksi sanaa',
      'with space': 'space',
      $dlr: 'val',
      _us_: 'underscore',
    }
    const result = queryObjectToString(o)
    expect(result).toMatchSnapshot()
  })
})
