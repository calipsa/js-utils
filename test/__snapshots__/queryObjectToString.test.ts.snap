// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sliceRotate should convert a non-english object correctly 1`] = `"%D0%BA%D0%BB%D1%8E%D1%87=%D0%B7%D0%BD%D0%B0%D1%87%D0%B5%D0%BD%D0%B8%D0%B5&anahtar=de%C4%9Fer&Schl%C3%BCssel=Werte"`;

exports[`sliceRotate should convert a normal object correctly 1`] = `"a=5&b=foo&tr=true&fs=false&trs=true&fss=false&emptystring=&d=2019-11-22T22%3A17%3A17.771Z&ud=undefined"`;

exports[`sliceRotate should convert a weird object correctly 1`] = `"twowords=kaksi+sanaa&with+space=space&%24dlr=val&_us_=underscore"`;

exports[`sliceRotate should convert an empty object correctly 1`] = `""`;

exports[`sliceRotate should convert an object with numeric keys correctly 1`] = `"0=zero&11=eleven"`;

exports[`sliceRotate should convert an object with various cases correctly 1`] = `"snake_case=val&camelCase=val&Capital=val&ALL_CAPS=val"`;
